### ISSUE TITLE
[Feature Request] Explicit `PYTHONOPTIMIZE` environment variable detection

### ISSUE THREAD

**Issue by jaanli** [beartype/beartype#341]
Minimal reproduction: 

test.py:

```
from beartype import beartype

@beartype
def test(x: int) -> None:
    print(x)

if __name__ == "__main__":
    test("test")
```

```
⬢ [Docker] ❯ python test.py 
Traceback (most recent call last):
  File "/home/ray/phare-sandbox/test.py", line 8, in <module>
    test("test")
  File "<@beartype(__main__.test) at 0x7ff83a8b8280>", line 29, in test
beartype.roar.BeartypeCallHintParamViolation: Function __main__.test() parameter x='test' violates type hint <class 'int'>, as str 'test' not instance of int.
```

```
⬢ [Docker] ❯ PYTHONOPTIMIZE=1 python test.py 
test
```

In a Jupyter notebook, this does not work to disable beartype:

![image](https://github.com/beartype/beartype/assets/5317244/247c7a62-9834-47c0-a3bf-4a81f2f8debc)

This is a blocker for projects that require JIT compilation such as with PyTorch compile / LLMs: https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html (the beartype decorator breaks compilation, which is needed to speed up LLM training and inference).

Loving beartype otherwise, it's saved me from many bugs -- rawr! 🐻

**Comment by leycec** [comment]
**Excellent issue!** Currently, *all* runtime type-checkers that I know of (including @beartype, `typeguard`, and Pydantic) just let Python handle `PYTHONOPTIMIZE`. That is, we don't explicitly detect `PYTHONOPTIMIZE`; we let Python itself detect `PYTHONOPTIMIZE` at startup, which then implicitly sets the `__debug__` global accordingly.

Clearly, this historical approach is overly simplistic. By letting Python handle everything, we fail to account for **dynamic changes to the `PYTHONOPTIMIZE` environment variable that a user manually makes** *after* Python has already started up.

Thankfully, this should be trivial for @beartype (and everybody else) to support. Thanks so much for the excellent issue that I can actually do something about, @jaanli. :smile:

That said, this...

> ...the beartype decorator breaks compilation

This saddens me! Would you mind opening a separate issue about that? That's actually the first I've heard of that. I mean, I'm not surprised. PyTorch is a huge thing. But I'd rather @beartype *not* break that huge thing if I can help it. Perhaps there's something that can be done on our end to improve @beartype **+** PyTorch integration.

Let's cogitate on this as I stroke my chin thoughtfully. :thinking:

**Comment by jaanli** [comment]
You're the bestest, thank you @leycec!! Will file a separate issue to reproduce :)

**Comment by leycec** [comment]
**Resolved by a342229.** The `@beartype` decorator now silently reduces to a noop (i.e., avoids decorating anything with runtime type-checking) when you manually set the `${PYTHONOPTIMIZED}` environment variable to a non-zero integer from within REPLs like Jupyter. Flex, @beartype!

Thanks so much for this *super* fun distraction, @jaanli. It's rare to receive a feature request that I can actually push out in an hour. Oh, if only they were all like this... :sweat_smile:
